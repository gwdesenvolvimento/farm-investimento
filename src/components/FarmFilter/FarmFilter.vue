<template>
    <div class="d-flex justify-center">
        <v-card class="filter">
            <v-card-title class="justify-center">
                Filtrar dados sobre um país 
            </v-card-title>
            <v-card-text>
                <v-row class="pr-16 pl-16">
                    <v-col cols="7" xs="12">
                        <v-text-field
                            label="Filtrar dados sobre um país "
                            placeholder="Digite o nome do País"
                            prepend-inner-icon="mdi-magnify"
                            color="primary"
                            v-model="countryName"
                            @change="filterCountry"
                        />
                    </v-col>
                    <v-col xs=12>
                        <v-radio-group 
                            v-model="optionOrder"
                            @change="orderCountries"
                        >
                            <v-radio
                                label="Ordenar A|Z"
                                value="ASC|COUNTRY"
                            />
                            <v-radio
                                label="Ordenar Z|A"
                                value="DESC|COUNTRY"
                            />
                            <v-radio
                                label="Maiores Totais de Casos"
                                value="ASC|TOTAIS"
                            />
                            <v-radio
                                label="Maiores Totais de Mortes"
                                value="ASC|MORTES"
                            />
                            <v-radio
                                label="Maiores Fatalidades"
                                value="ASC|FATALIDADES"
                            />
                        </v-radio-group>
                    </v-col>
                </v-row>
            </v-card-text>
        </v-card>
    </div>
</template>
<style scoped>
    .filter {
        width: 70%;
        margin-top: -80px;
        border: 1px solid var(--light);
    }

    .v-card__title  {
        font-size: 1.5rem;
        font-weight: 700;
        color: var(--gray);
    }

</style>
<script>
export default {
    name: 'FarmFilter',
    data: () => ({
        countryName: '',
        optionOrder: 'ASC'
    }),
    methods: {
        filterCountry() {
            this.$store.dispatch('filterLstCovidCases', this.countryName);
        },
        orderCountries() {
            this.$store.dispatch('orderListOfCountries', this.optionOrder)
        }
    }
};
</script>