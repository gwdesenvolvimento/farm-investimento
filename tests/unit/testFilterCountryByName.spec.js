import { createLocalVue, mount } from '@vue/test-utils';
import FarmFilter from '../../src/components/FarmFilter/FarmFilter.vue';
import Vuetify from 'vuetify';
import Vuex from 'vuex';

describe('Testando componente FarmFilter', () => {

  let localVue;
  let vuetify;
  let wrapper;
  let store;
  
  const storeCovid = {
    state: {
      lstCovidCases: [
        { Country: 'Spain', cases: 200000, deaths: 10000, caseFatalityRate: 0.05 },
        { Country: 'Argentina', cases: 500000, deaths: 25000, caseFatalityRate: 0.05 },
        { Country: 'Brazil', cases: 100000, deaths: 5000, caseFatalityRate: 0.05 },
      ],
      lstFilteredCovidCases: [
        { Country: 'Spain', cases: 200000, deaths: 10000, caseFatalityRate: 0.05 },
        { Country: 'Argentina', cases: 500000, deaths: 25000, caseFatalityRate: 0.05 },
        { Country: 'Brazil', cases: 100000, deaths: 5000, caseFatalityRate: 0.05 },
      ],
      showLoading: false,
      apiMessage: '',
      isCachingInProgress: false,
    },
    getters: {
      lstFilteredCovidCases: state => state.lstFilteredCovidCases
    },
    mutations: {
      filterLstCovidCases(state, payload) {
        state.showLoading = true
        
        if (payload !== "") {
          state.lstFilteredCovidCases = state.lstCovidCases.filter(item => item.Country.indexOf(payload) !== -1)
          state.showLoading = !(state.lstFilteredCovidCases.length > 0)
          state.apiMessage = ''
          return 
        }
        
        state.lstFilteredCovidCases = state.lstCovidCases
        state.showLoading = !(state.lstFilteredCovidCases.length > 0)
        state.apiMessage = ''
        state.isCachingInProgress = false
      },
      orderListOfCountries(state, direction) {
          console.log(direction)
          let arrDirection = direction.split('|')
          direction = arrDirection[0]
          let field = arrDirection[1]
          
          if (field === "COUNTRY") {
            
            switch (direction) {
              case "ASC": state.lstFilteredCovidCases = state.lstFilteredCovidCases.sort((a, b) => a.Country.localeCompare(b.Country)); break;
              default: state.lstFilteredCovidCases = state.lstFilteredCovidCases.sort((a, b) => b.Country.localeCompare(a.Country)); break;
            }
          }

          switch (field) {
              case "TOTAIS": state.lstFilteredCovidCases = state.lstFilteredCovidCases.sort((a, b) => b.TotalConfirmed - a.TotalConfirmed); break;
              case "MORTES": state.lstFilteredCovidCases = state.lstFilteredCovidCases.sort((a, b) => b.TotalDeaths - a.TotalDeaths); break;
              case "FATALIDADES": state.lstFilteredCovidCases = state.lstFilteredCovidCases.sort((a, b) => b.fatalidadePercentual - a.fatalidadePercentual); break;
          }
      }
    },
    actions: {
      filterLstCovidCases({ commit }, countryName) {
        commit('filterLstCovidCases', countryName)
      },
      orderListOfCountries({ commit }, direction) {
        commit('orderListOfCountries', direction)
      }
    }
  }

  beforeEach(() => {
    localVue = createLocalVue();
    localVue.use(Vuex)
    store = new Vuex.Store(storeCovid);  
    vuetify = new Vuetify()
    localVue.use(vuetify)
    
    wrapper = mount(FarmFilter, {
      localVue,
      vuetify,
      store,
      attachTo: document.body,
    });
  });

  afterEach(() => {
    wrapper.destroy();
  });

  test('Verifica se uma instancia do Vue foi criada', () => {
    expect(wrapper.vm).toBeTruthy();
  });

  test('Verifica se Renderizou corretamente', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  test('Verifica se o Filtro está funcionando corretamente.', async () => {
    const filterInput = wrapper.find('[data-testid="input-field"]');
    await wrapper.setData({ countryName: "Brazil" });
    filterInput.trigger('keydown');
    const filteredCountries = store.state.lstFilteredCovidCases;
    expect(filteredCountries.length).toBe(1);
    expect(filteredCountries[0].Country).toBe('Brazil');
  });

  test('Verifica se o Filtro está voltando com todas os países.', async () => {
    const filterInput = wrapper.find('[data-testid="input-field"]');
    await wrapper.setData({ countryName: "" });
    filterInput.trigger('keydown');
    const filteredCountries = store.state.lstFilteredCovidCases;
    expect(filteredCountries.length).toBe(3);
    expect(filteredCountries[0].Country).toBe('Spain');
  });

});