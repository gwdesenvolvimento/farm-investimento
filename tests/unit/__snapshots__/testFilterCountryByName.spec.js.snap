// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testando componente FarmFilter Verifica se Renderizou corretamente 1`] = `
<div
  class="d-flex justify-center"
>
  <div
    class="filter v-card v-sheet theme--light"
  >
    <div
      class="v-card__title justify-center"
    >
      
            Filtrar dados sobre um país 
        
    </div>
     
    <div
      class="v-card__text"
    >
      <div
        class="row pr-md-16 pl-md-16"
      >
        <div
          class="col-md-7 col-12 col"
        >
          <div
            class="v-input theme--light v-text-field v-text-field--placeholder"
          >
            <div
              class="v-input__control"
            >
              <div
                class="v-input__slot"
              >
                <div
                  class="v-input__prepend-inner"
                >
                  <div
                    class="v-input__icon v-input__icon--prepend-inner"
                  >
                    <i
                      aria-hidden="true"
                      class="v-icon notranslate mdi mdi-magnify theme--light"
                    />
                  </div>
                </div>
                <div
                  class="v-text-field__slot"
                >
                  <label
                    class="v-label theme--light"
                    for="input-27"
                    style="left: 0px; position: absolute;"
                  >
                    Filtrar dados sobre um país 
                  </label>
                  <input
                    data-testid="input-field"
                    id="input-27"
                    type="text"
                  />
                </div>
              </div>
              <div
                class="v-text-field__details"
              >
                <div
                  class="v-messages theme--light"
                >
                  <transition-group-stub
                    class="v-messages__wrapper"
                    name="message-transition"
                    tag="div"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
         
        <div
          class="col"
          xs="12"
        >
          <div
            class="v-input v-input--is-label-active v-input--is-dirty theme--light v-input--selection-controls v-input--radio-group v-input--radio-group--column"
          >
            <div
              class="v-input__control"
            >
              <div
                class="v-input__slot"
                style="height: auto;"
              >
                <div
                  aria-labelledby="input-31"
                  class="v-input--radio-group__input"
                  role="radiogroup"
                >
                  <legend
                    class="v-label theme--light"
                    id="input-31"
                    style="left: 0px; position: relative;"
                  >
                    Ordenação
                  </legend>
                  <div
                    class="v-radio theme--light v-item--active"
                  >
                    <div
                      class="v-input--selection-controls__input"
                    >
                      <i
                        aria-hidden="true"
                        class="v-icon notranslate mdi mdi-radiobox-marked theme--light primary--text"
                      />
                      <input
                        aria-checked="true"
                        id="input-32"
                        name="radio-31"
                        role="radio"
                        type="radio"
                        value="ASC|COUNTRY"
                      />
                      <div
                        class="v-input--selection-controls__ripple primary--text"
                      />
                    </div>
                    <label
                      class="v-label theme--light"
                      for="input-32"
                      style="left: 0px; position: relative;"
                    >
                      A|Z
                    </label>
                  </div>
                   
                  <div
                    class="v-radio theme--light"
                  >
                    <div
                      class="v-input--selection-controls__input"
                    >
                      <i
                        aria-hidden="true"
                        class="v-icon notranslate mdi mdi-radiobox-blank theme--light"
                      />
                      <input
                        aria-checked="false"
                        id="input-34"
                        name="radio-31"
                        role="radio"
                        type="radio"
                        value="DESC|COUNTRY"
                      />
                      <div
                        class="v-input--selection-controls__ripple"
                      />
                    </div>
                    <label
                      class="v-label theme--light"
                      for="input-34"
                      style="left: 0px; position: relative;"
                    >
                      Z|A
                    </label>
                  </div>
                   
                  <div
                    class="v-radio theme--light"
                  >
                    <div
                      class="v-input--selection-controls__input"
                    >
                      <i
                        aria-hidden="true"
                        class="v-icon notranslate mdi mdi-radiobox-blank theme--light"
                      />
                      <input
                        aria-checked="false"
                        id="input-36"
                        name="radio-31"
                        role="radio"
                        type="radio"
                        value="ASC|TOTAIS"
                      />
                      <div
                        class="v-input--selection-controls__ripple"
                      />
                    </div>
                    <label
                      class="v-label theme--light"
                      for="input-36"
                      style="left: 0px; position: relative;"
                    >
                      Totais de Casos
                    </label>
                  </div>
                   
                  <div
                    class="v-radio theme--light"
                  >
                    <div
                      class="v-input--selection-controls__input"
                    >
                      <i
                        aria-hidden="true"
                        class="v-icon notranslate mdi mdi-radiobox-blank theme--light"
                      />
                      <input
                        aria-checked="false"
                        id="input-38"
                        name="radio-31"
                        role="radio"
                        type="radio"
                        value="ASC|MORTES"
                      />
                      <div
                        class="v-input--selection-controls__ripple"
                      />
                    </div>
                    <label
                      class="v-label theme--light"
                      for="input-38"
                      style="left: 0px; position: relative;"
                    >
                      Totais de Mortes
                    </label>
                  </div>
                   
                  <div
                    class="v-radio theme--light"
                  >
                    <div
                      class="v-input--selection-controls__input"
                    >
                      <i
                        aria-hidden="true"
                        class="v-icon notranslate mdi mdi-radiobox-blank theme--light"
                      />
                      <input
                        aria-checked="false"
                        id="input-40"
                        name="radio-31"
                        role="radio"
                        type="radio"
                        value="ASC|FATALIDADES"
                      />
                      <div
                        class="v-input--selection-controls__ripple"
                      />
                    </div>
                    <label
                      class="v-label theme--light"
                      for="input-40"
                      style="left: 0px; position: relative;"
                    >
                      Fatalidades
                    </label>
                  </div>
                </div>
              </div>
              <div
                class="v-messages theme--light"
              >
                <transition-group-stub
                  class="v-messages__wrapper"
                  name="message-transition"
                  tag="div"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
